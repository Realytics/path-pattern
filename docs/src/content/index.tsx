import * as React from 'react';
import { PathPatternDoc } from './PathPatternDoc';
import { Row } from '../components/Row';

const logo = require('../../assets/logo.svg');

export interface ContentProps {}

export const Content: React.SFC<ContentProps> = () => {
  return (
    <div className="content">
      <Row>
        <h1>
          <img src={logo} alt="PathPattern" className="logo" />
        </h1>
      </Row>
      <PathPatternDoc />
    </div>
  );
};
