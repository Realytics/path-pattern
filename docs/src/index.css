html {
  font-size: 55%;
  height: 100%;
}

body {
  min-height: 100vh;
}

#root {
  min-height: 100vh;
}

.content {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  /* margin-left: 20px; */
}

pre {
  border-left: none;
  margin: 0;
}

pre > code {
  padding: 0.5rem 1rem;
  tab-size: 2;
  background: none;
}

code {
  font-family: 'Fira Code', Consolas, Monaco, Andale Mono, Ubuntu Mono, monospace;
}

p {
  margin-bottom: 1rem;
}

ul {
  padding-left: 3rem;
}

li {
  list-style-type: none;
  position: relative;
  margin-bottom: 0.3em;
}

li:before {
  content: 'â€¢';
  display: inline-block;
  position: absolute;
  left: -1em;
  color: #00c7ba;
}

h1 {
  margin-top: 2rem;
}

h2 {
  font-weight: 500;
}

.logo {
  width: 90%;
}

.doc-type {
  font-weight: 600;
  font-size: 0.7em;
  text-transform: uppercase;
  margin-right: 10px;
}

.title-type {
  font-weight: 100;
  text-transform: uppercase;
  background: #1e88e5;
  color: white;
  font-size: 1.4rem;
  vertical-align: middle;
  margin-right: 6px;
  margin-left: 10px;
  margin-bottom: 3px;
  letter-spacing: 0;
  display: inline-block;
}

.method-title-type {
  font-weight: 100;
  text-transform: uppercase;
  background: #1e88e5;
  color: white;
  font-size: 60%;
  margin-left: 10px;
  padding: 0.2rem 0.5rem;
  line-height: 1.5;
  display: inline-block;
  vertical-align: middle;
  padding-top: 0.4rem;
  letter-spacing: 0;
}

.doc-row {
  display: flex;
  flex-direction: row;
}

.doc-row > * {
  flex-shrink: 0;
}

.code {
  display: flex;
  flex-direction: column;
  flex: 1;
  background: #282c34;
}

.code-row {
  display: flex;
  flex-direction: row;
  max-width: 800px;
}

.txt {
  padding: 0 2rem;
  padding-top: 1rem;
  width: 450px;
}

.txt pre {
  border-radius: 4px;
}

.output {
  flex-direction: column;
  align-items: stretch;
  justify-content: stretch;
  flex: 1;
  flex-shrink: 0;
  overflow: hidden;
}

.output.toggle {
  cursor: pointer;
}

.output > pre > code {
  opacity: 0.5;
  transition-duration: 0.3s;
}

.output > pre > code:hover {
  opacity: 1;
}

.codesource {
  flex-direction: column;
  align-items: stretch;
  justify-content: stretch;
  width: 50%;
  flex-shrink: 0;
}

.hljs-doctag,
.hljs-keyword,
.hljs-formula {
  color: #cda6d8;
}

.hljs-string,
.hljs-regexp,
.hljs-addition,
.hljs-attribute,
.hljs-meta-string {
  color: #b4cea1;
}

@media (max-width: 1080px) {
  .code-row {
    flex-direction: column;
  }
  .codesource {
    width: auto;
  }

  .output > pre > code {
    opacity: 1;
    transition-duration: 0.3s;
  }
  .output {
    position: relative;
    margin: 1rem;
    margin-top: 0;
  }
  .output > pre {
    background: rgba(255, 255, 255, 0.1);
  }
  /* .output::before {
    content: '';
    position: absolute;
    top: 0;
    left: 10px;
    right: 10px;
    height: 1px;
    background: rgba(255, 255, 255, 0.1);
  } */
}

@media (max-width: 800px) {
  .doc-row {
    flex-direction: column;
  }
  .code-row {
    padding-top: 0.5rem;
  }
  .code {
    border-radius: 4px;
    overflow: hidden;
    margin-bottom: 2rem;
  }
  .code.code-empty {
    display: none;
  }
  .content {
    margin-right: 20px;
    margin-left: 20px;
  }
  .txt {
    width: auto;
    padding: 0;
  }
  .txt code {
    white-space: normal;
  }
}
