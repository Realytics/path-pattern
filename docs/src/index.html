<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Path-Pattern documentation</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300italic,500,500italic,600,600italic,700,700italic">
  <link rel="stylesheet" href="https://cdn.rawgit.com/tonsky/FiraCode/1.205/distr/fira_code.css">
  <link rel="stylesheet" href="../node_modules/highlight.js/styles/atom-one-dark.css">
  <link rel="stylesheet" href="../node_modules/normalize.css/normalize.css">
  <link rel="stylesheet" href="../node_modules/milligram/dist/milligram.min.css">
  <link rel="stylesheet" href="./index.css">
</head>

<body>
  <div id="root"></div>
  <script src="./index.tsx"></script>
</body>

</html>