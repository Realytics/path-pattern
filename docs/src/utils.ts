import * as hljs from 'highlight.js';
import * as MarkdownIt from 'markdown-it';

export const mdIt: MarkdownIt.MarkdownIt = new MarkdownIt({
  html: true,
  linkify: true,
  highlight: (str, lang) => {
    if (lang && hljs.getLanguage(lang)) {
      try {
        return '<pre class="hljs"><code>' + hljs.highlight(lang, str, true).value + '</code></pre>';
      } catch (error) {
        // do nothing
        console.error(error);
      }
    }

    return '<pre class="hljs"><code>' + mdIt.utils.escapeHtml(str) + '</code></pre>';
  },
});

export type CodeObject<T> = {
  source: string;
  // cb: () => T;
  output: T;
  key: string;
};

export function code<T>(codeContent: any, id: string, cb: () => T): CodeObject<T> {
  return {
    source: codeContent[id],
    // cb,
    output: cb(),
    key: id,
  };
}

export function codeParams<T, P>(codeContent: any, id: string, cb: (params: P) => T, params: P): CodeObject<T> {
  return {
    source: codeContent[id],
    // cb: () => cb(params),
    output: cb(params),
    key: id,
  };
}
