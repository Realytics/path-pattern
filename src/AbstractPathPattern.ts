
import { Location } from 'history';
import { normalizePathPattern } from './normalizePathPattern';
import { Match } from './Match';
import { Matchable, MatchOptions } from './Matchable';
import { PatternAware } from './PatternAware';

export abstract class AbstractPathPattern<Params> implements Matchable<Params>, PatternAware {

  protected pattern: string;

  constructor(pattern: string) {
    this.pattern = normalizePathPattern(pattern);

    this.matchAdvanced = this.matchAdvanced.bind(this);
    this.match = this.match.bind(this);
    this.matchExact = this.matchExact.bind(this);
    this.matchStrict = this.matchStrict.bind(this);
  }

  getPattern(): string {
    return this.pattern;
  }

  abstract matchAdvanced(options: MatchOptions): (location: Location) => Match<Params>;
  abstract match(location: Location): Match<Params>;
  abstract matchExact(location: Location): Match<Params>;
  abstract matchStrict(location: Location): Match<Params>;
}
