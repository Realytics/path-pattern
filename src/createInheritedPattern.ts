import { normalizePathPattern } from './normalizePathPattern';
import { createPattern } from './createPattern';
import { PathPattern } from './types';

export function createInheritedPattern<ParentParams = any, Params = any>(
  parentPathPattern: PathPattern<ParentParams>,
  pattern: string
): PathPattern<ParentParams & Params> {
  return createPattern(parentPathPattern.getPattern() + normalizePathPattern(pattern));
}
