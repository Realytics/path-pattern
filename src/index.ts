export { PathPattern, PathPatternWithParams } from './PathPattern';
export { RelativePathPattern } from './RelativePathPattern';
export { normalizePathPattern } from './normalizePathPattern';
export { matchOneOf } from './matchOneOf';
export { Match, MatchSuccess, MatchOptions, Matcher, Pattern, PatternWithParams } from './types';
