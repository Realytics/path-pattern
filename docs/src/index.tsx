import * as React from 'react';
import { Content } from './content';
import { PathPattern } from '../src';
import { render } from 'react-dom';

render(<Content />, document.getElementById('root'));
