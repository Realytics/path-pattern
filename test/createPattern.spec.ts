import { Location } from 'history';
import { createPattern } from '../src';

describe(`createPattern('/home')`, () => {
  const pattern = createPattern('/home');
  it(`match '/home'`, () => {
    expect(pattern.match('/home')).toBeTruthy();
  });
  it(`match '/home/user'`, () => {
    expect(pattern.match('/home/user')).toBeTruthy();
  });
  it(`does not match '/welcome'`, () => {
    expect(pattern.match('/welcome')).toBeFalsy();
  });
  it(`does not match '/'`, () => {
    expect(pattern.match('/')).toBeFalsy();
  });
  it(`does not match with no location or no pathname`, () => {
    expect(pattern.match(null)).toBeFalsy();
    expect(pattern.match(undefined)).toBeFalsy();
  });
  it(`compile to '/home'`, () => {
    expect(pattern.compile({})).toEqual('/home');
  });
});

describe(`createPattern('/home', { exact: true })`, () => {
  const pattern = createPattern('/home');
  it(`match '/home'`, () => {
    expect(pattern.matchExact('/home')).toBeTruthy();
  });
  it(`match '/home/'`, () => {
    expect(pattern.matchExact('/home/')).toBeTruthy();
  });
  it(`does not match '/home/user'`, () => {
    expect(pattern.matchExact('/home/user')).toBeFalsy();
  });
  it(`does not match '/welcome'`, () => {
    expect(pattern.matchExact('/welcome')).toBeFalsy();
  });
  it(`does not match '/'`, () => {
    expect(pattern.matchExact('/')).toBeFalsy();
  });
  it(`get getPattern`, () => {
    expect(pattern.getPattern()).toEqual('/home');
  });
});

describe(`createPattern('/home/', { exact: true, strict: true })`, () => {
  const pattern = createPattern('/home/');
  it(`match '/home/'`, () => {
    expect(pattern.matchStrict('/home/')).toBeTruthy();
  });
  it(`does not match '/home'`, () => {
    expect(pattern.matchStrict('/home')).toBeFalsy();
  });
  it(`does not match '/home/user'`, () => {
    expect(pattern.matchStrict('/home/user')).toBeFalsy();
  });
  it(`does not match '/welcome'`, () => {
    expect(pattern.matchStrict('/welcome')).toBeFalsy();
  });
  it(`does not match '/'`, () => {
    expect(pattern.matchStrict('/')).toBeFalsy();
  });
  it(`compile to '/home'`, () => {
    expect(pattern.compile({})).toEqual('/home/');
  });
  it(`get getPattern`, () => {
    expect(pattern.getPattern()).toEqual('/home/');
  });
});

// create twice the same path for coverage
describe(`createPattern('/home/', { exact: true, strict: true }) bis`, () => {
  const pattern = createPattern('/home/');
  it(`match '/home/'`, () => {
    expect(pattern.matchStrict('/home/')).toBeTruthy();
  });
  it(`does not match '/home'`, () => {
    expect(pattern.matchStrict('/home')).toBeFalsy();
  });
  it(`compile to '/home'`, () => {
    expect(pattern.compile({})).toEqual('/home/');
  });
});

describe(`createPattern('/home/', { strict: true })`, () => {
  const pattern = createPattern('/home/');
  it(`match '/home/'`, () => {
    expect(pattern.matchStrict('/home/')).toBeTruthy();
  });
  it(`does not match '/home'`, () => {
    expect(pattern.matchStrict('/home')).toBeFalsy();
  });
  it(`match '/home/user'`, () => {
    expect(pattern.matchStrict('/home/user')).toBeFalsy();
  });
  it(`does not match '/welcome'`, () => {
    expect(pattern.matchStrict('/welcome')).toBeFalsy();
  });
  it(`does not match '/'`, () => {
    expect(pattern.matchStrict('/')).toBeFalsy();
  });
});

describe(`createPattern('/user/:user')`, () => {
  const pattern = createPattern('/user/:user');
  it(`does not match '/'`, () => {
    expect(pattern.match('/')).toBeFalsy();
  });
  it(`does not match '/user'`, () => {
    expect(pattern.match('/user')).toBeFalsy();
  });
  it(`does not match '/user/'`, () => {
    expect(pattern.match('/user/')).toBeFalsy();
  });
  it(`match '/user/john'`, () => {
    expect(pattern.match('/user/john')).toBeTruthy();
  });
  it(`match '/user/john'`, () => {
    expect(pattern.match('/user/john')).toMatchSnapshot();
  });
  it(`compile with user jane`, () => {
    expect(pattern.compile({ user: 'jane' })).toEqual('/user/jane');
  });
  it(`compile with user john`, () => {
    expect(pattern.compile({ user: 'john' })).toEqual('/user/john');
  });
  it(`get getPattern`, () => {
    expect(pattern.getPattern()).toEqual('/user/:user');
  });
});

describe(`createPattern('/user/:userName')`, () => {
  const pattern = createPattern('/user/:userName');
  it(`compile with user userName`, () => {
    expect(pattern.compile({ userName: 'jane' })).toEqual('/user/jane');
  });
  it(`does not compile with wrong param`, () => {
    expect(() => pattern.compile({ yolo: 'jane' })).toThrowErrorMatchingSnapshot();
  });
  it(`compile with user username`, () => {
    expect(() => pattern.compile({ username: 'john' })).toThrowErrorMatchingSnapshot();
  });
});

describe(`createPattern('/')`, () => {
  const pattern = createPattern('/');
  it(`does not match ''`, () => {
    expect(pattern.match('')).toBeFalsy();
    expect(pattern.matchAdvanced({})('')).toBeFalsy();
  });
  it(`match '/'`, () => {
    expect(pattern.match('/')).toMatchSnapshot();
  });
});

describe(`createPattern('home')`, () => {
  const pattern = createPattern('home');
  it(`match '/home/'`, () => {
    expect(pattern.match('/home/')).toBeTruthy();
  });
  it(`match '/home'`, () => {
    expect(pattern.match('/home')).toBeTruthy();
  });
  it(`match '/home/user'`, () => {
    expect(pattern.match('/home/user')).toBeTruthy();
  });
  it(`does not match '/welcome'`, () => {
    expect(pattern.match('/welcome')).toBeFalsy();
  });
  it(`does not match '/'`, () => {
    expect(pattern.match('/')).toBeFalsy();
  });
  it(`get getPattern`, () => {
    expect(pattern.getPattern()).toEqual('/home');
  });
});
