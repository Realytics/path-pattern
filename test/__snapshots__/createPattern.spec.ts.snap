// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPattern('/user/:user') match '/user/john' 1`] = `
Object {
  "isExact": true,
  "params": Object {
    "user": "john",
  },
  "path": "/user/:user",
  "url": "/user/john",
}
`;

exports[`createPattern('/user/:userName') compile with user username 1`] = `"Expected \\"userName\\" to be a string"`;

exports[`createPattern('/user/:userName') does not compile with wrong param 1`] = `"Expected \\"userName\\" to be a string"`;
